File: 90.subsets-ii.java

// @leet imports start
import java.util.*;
import java.math.*;
// @leet imports end

// @leet start
class Solution {
    List<List<Integer>> ans;
    int[] nums;

    public List<List<Integer>> subsetsWithDup(int[] nums) {
        this.ans = new ArrayList<>();
        this.nums = nums;

        List<Integer> curr = new ArrayList<>();
        Arrays.sort(nums);
        dfs(0, curr);
        return ans;

    }

    private void dfs(int start, List<Integer> curr) {
        List<Integer> copy = new ArrayList<>(curr);
        ans.add(copy);

        for (int i = start; i < nums.length; i++) {
            if (i > start && nums[i] == nums[i - 1]) {
                continue;
            }
            curr.add(nums[i]);
            dfs(i + 1, curr);
            curr.removeLast();
        }
    }
}
// @leet end

File: 90.subsets-ii.py
from typing import List


# @leet start
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        ans = []
        nums.sort()  # so you can skip
        n = len(nums)

        def dfs(first: int, curr: list[int]):
            ans.append(curr[:])

            for i in range(first, n):
                # subset2 specific things
                if i > first and nums[i] == nums[i - 1]:
                    continue
                # end
                curr.append(nums[i])
                dfs(i + 1, curr)
                curr.pop()

        dfs(0, [])
        return ans


# @leet end
